### PR TITLE
Ds 3049

### PR DESCRIPTION
**Pull request by chschan**

### CODE REVIEW

File: tests/testthat/test-correspondenceanalysis.R [hunk 1]
```diff
@@ -546,3 +588,78 @@ test_that("Output contains the right class for extension buttons", {
...
+        FilteredProportion = 0), questiontypes = "PickAnyGrid", span = list(
+        rows = structure(list(c("Coke", "Diet Coke", "Coke Zero",
+        "Pepsi", "Diet Pepsi", "Pepsi Max", "None of these", "NET"
+        )), class = "data.frame", .Names = "", row.names = c(NA,
+        8L)), columns = structure(list(c("Feminine", "Health-conscious",
+        "Innocent", "Older", "Open to new experiences", "Rebellious",
+        "Sleepy", "Traditional", "Weight-conscious", "NET")), class = "data.frame", .Names = "", row.names = c(NA,
+        10L))), name = "table.Q5", questions = c("Q5", "SUMMARY"))
+
+    b.size <- structure(c(6.42201834862385, 57.4923547400612, 22.3241590214067,
```
Comment (JustinCCYap) [inline tests/testthat/test-correspondenceanalysis.R]: Is this exactly the same as `x`? If so, why is it defined again?

File: tests/testthat/test-correspondenceanalysis.R [hunk 2]
```diff
@@ -166,10 +188,29 @@ test_that("Bubble charts",
...
+                expect_warning(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart",
+                    bubble.size = c(Extra = 1, More = 2, bsizes)), "Bubble sizes for 'Extra', 'More' were ignored.")
+                expect_error(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart", bubble.size = bsizes[-1]),
+                    "Bubble sizes: Missing values for 'Coke'")
+                expect_error(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart",
+                    bubble.size = c(bsizes[-1], 'coke ' = 2)), NA)
+
                 names(bsizes)[2] = "Dog"
-                expect_error(print(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart", bubble.size = bsizes)), "The bubble sizes must contain the same names")
+                expect_error(print(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart", bubble.size = bsizes)), "Bubble sizes: Missing values for 'V'")
```
Comment (JustinCCYap) [inline tests/testthat/test-correspondenceanalysis.R]: I think the error message could be improved. Tim said in the jira: "To use a Bubble Chart, the table of Bubble sizes needs to include all the row categories used in the analysis. The following are missing: Coke, Pepsi"

File: R/correspondenceanalysis.R [hunk 3]
```diff
@@ -238,32 +233,25 @@ CorrespondenceAnalysis = function(x,
...
+                bubble.size <- bubble.size[,1,drop = FALSE]
+            }
+            bubble.size <- TidyTabularData(bubble.size,
+                row.names.to.remove = row.names.to.remove)
+            b.orig.names <- rownames(bubble.size)
+            bubble.size <- MatchTable(bubble.size, ref.names = rownames(x),
+                x.table.name = "Bubble sizes")
+            extra.bnames <- setdiff(b.orig.names, rownames(bubble.size))
+            if (length(extra.bnames) > 0)
+                warning("Bubble sizes for '", paste(extra.bnames, collapse = "', '"),
```
Comment (JustinCCYap) [inline R/correspondenceanalysis.R:271]: Need to deal with the singular case too where there is only one element in `extra.bnames`

Comment (JustinCCYap) [reply R/correspondenceanalysis.R:271]: And mention why it was ignored: "because they do not appear in the input data."

File: tests/testthat/test-correspondenceanalysis.R (hunk 2 above)
Comment (JustinCCYap) [reply tests/testthat/test-correspondenceanalysis.R]: Remember to deal with singular/plural cases.

File: tests/testthat/test-correspondenceanalysis.R [hunk 4]
```diff
@@ -166,10 +188,29 @@ test_that("Bubble charts",
...
                 names(bsizes)[2] = names(bsizes)[1]
-                expect_error(print(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart", bubble.size = bsizes)), "There are duplicate names in bubble sizes")
+                expect_warning(expect_error(print(CorrespondenceAnalysis(x.with.labels, output = "Bubble Chart", bubble.size = bsizes)),
+                        "Bubble sizes: Missing values for 'V'"), "duplicate names")
```
Comment (JustinCCYap) [inline tests/testthat/test-correspondenceanalysis.R]: The warning message for this needs to be improved as well to be a full sentence. Perhaps: "The table of Bubble sizes contains duplicate row labels for 'Coke'. The value from the first duplicate was used."

File: tests/testthat/test-correspondenceanalysis.R [hunk 5]
```diff
@@ -546,3 +588,78 @@ test_that("Output contains the right class for extension buttons", {
...
+        FilteredProportion = 0), questiontypes = "PickAnyGrid", span = list(
+        rows = structure(list(c("Coke", "Diet Coke", "Coke Zero",
+        "Pepsi", "Diet Pepsi", "Pepsi Max", "None of these", "NET"
+        )), class = "data.frame", .Names = "", row.names = c(NA,
+        8L)), columns = structure(list(c("Feminine", "Health-conscious",
+        "Innocent", "Older", "Open to new experiences", "Rebellious",
+        "Sleepy", "Traditional", "Weight-conscious", "NET")), class = "data.frame", .Names = "", row.names = c(NA,
+        10L))), name = "table.Q5", questions = c("Q5", "SUMMARY"))
+
+    res <- CorrespondenceAnalysis(x, output = "Bubble Chart", bubble.size = b.size[,1])
```
Comment (JustinCCYap) [inline tests/testthat/test-correspondenceanalysis.R]: Should the index `[,1]` be removed so that you can test the warning "Only first column of bubble sizes is used."

File: R/correspondenceanalysis.R [hunk 6]
```diff
@@ -238,32 +233,25 @@ CorrespondenceAnalysis = function(x,
...
-         #       }
-            # Sorting bubble sizes to match the row names of the table.
-            #order = match(rownames(x), bubble.names)
-            #if (sum(order, na.rm = TRUE) != sum(1:length(order)))
-            #    stop("The bubble sizes must contain the same names as in the rows of the input data table: ",
-            #         paste0(paste0(table.names, ":", bubble.names), collapse = ", "), ".")
-            #bubble.size = bubble.size[order]
+            bubble.size <- as.matrix(bubble.size)
+            if (is.null(rownames(bubble.size)))
+                stop("The bubble sizes need to be named to match the row labels of the input data.")
```
Comment (JustinCCYap) [inline R/correspondenceanalysis.R]: Reword this to be more consistent in terminology with the missing values error (see below).

File: R/correspondenceanalysis.R [hunk 7]
```diff
@@ -238,32 +233,25 @@ CorrespondenceAnalysis = function(x,
...
+            if (NCOL(bubble.size) > 1)
+            {
+                warning("Only first column of bubble sizes is used.")
```
Comment (JustinCCYap) [inline R/correspondenceanalysis.R]: Reword this to mention: "The table of bubble sizes contains more than one column. Only first column of bubble sizes was used."

File: tests/testthat/test-correspondenceanalysis.R (hunk 1 above)
Comment (chschan) [reply tests/testthat/test-correspondenceanalysis.R]: Sorry, this was work in progress that I forgot to remove. I've deleted this section now in 35da5f2

File: R/correspondenceanalysis.R (hunk 7 above)
Comment (chschan) [reply R/correspondenceanalysis.R]: Fixed in 35da5f2

File: tests/testthat/test-correspondenceanalysis.R (hunk 5 above)
Comment (chschan) [reply tests/testthat/test-correspondenceanalysis.R]: This warning is now tested in line 221 in 35da5f2

File: tests/testthat/test-correspondenceanalysis.R (hunk 4 above)
Comment (chschan) [reply tests/testthat/test-correspondenceanalysis.R]: Fixed using flipChartBasics@DS-3049-2

File: R/correspondenceanalysis.R (hunk 6 above)
Comment (chschan) [reply R/correspondenceanalysis.R]: Fixed in 35da5f2

File: R/correspondenceanalysis.R (hunk 3 above)
Comment (chschan) [reply R/correspondenceanalysis.R:271]: Fixed in 35da5f2

File: R/correspondenceanalysis.R [hunk 8]
```diff
@@ -238,32 +234,36 @@ CorrespondenceAnalysis = function(x,
...
-         #       }
-            # Sorting bubble sizes to match the row names of the table.
-            #order = match(rownames(x), bubble.names)
-            #if (sum(order, na.rm = TRUE) != sum(1:length(order)))
-            #    stop("The bubble sizes must contain the same names as in the rows of the input data table: ",
-            #         paste0(paste0(table.names, ":", bubble.names), collapse = ", "), ".")
-            #bubble.size = bubble.size[order]
+            bubble.size <- as.matrix(bubble.size)
+            if (is.null(rownames(bubble.size)))
+                stop("The table of bubble sizes need to be named to match the row labels of the input data.")
```
Comment (JustinCCYap) [inline R/correspondenceanalysis.R]: Sorry as it is my wording but could you replace "row labels of the input data" with "row labels used in the analysis" to be more consistent?

File: R/correspondenceanalysis.R [hunk 9]
```diff
@@ -238,32 +234,36 @@ CorrespondenceAnalysis = function(x,
...
+            }
+            bubble.size <- TidyTabularData(bubble.size,
+                row.names.to.remove = row.names.to.remove)
+            b.orig.names <- rownames(bubble.size)
+
+            # Modify messages from MatchTable to inform users
+            # that the problem is in bubble.size
+            bubble.size <- InterceptExceptions(MatchTable(bubble.size, ref.names = rownames(x)),
+                error.handler = function(e){
+                    stop("To use a bubble chart, the table of bubble sizes needs to include all the row labels used in the analysis. ", e$message)
```
Comment (JustinCCYap) [inline R/correspondenceanalysis.R]: Should you check the error message to see if it actually is due to missing values? If it isn't, just rethrow the error without appending anything? Same with the warning below.

File: R/correspondenceanalysis.R (hunk 8 above)
Comment (chschan) [reply R/correspondenceanalysis.R]: Thanks for checking the consistency. Fixed in 52ac5ae

File: R/correspondenceanalysis.R (hunk 9 above)
Comment (chschan) [reply R/correspondenceanalysis.R]: Fixed in 52ac5ae

**Review by JustinCCYap — APPROVED** [review]
